manticore/core/cpu/binja.py

```diff
@@ -0,0 +1,1188 @@
+            # mask off the value that will be replaced
+            mask = (1 << ilreg.size * 8) - 1
+            full_mask = (1 << ilreg_full.size * 8) - 1
+            reg_bits = mask << (ilreg.offset * 8)
+            full_width_reg_value &= full_mask ^ reg_bits
+            full_width_reg_value |= value << ilreg.offset * 8
+        elif ilreg.extend == ImplicitRegisterExtend.ZeroExtendToFullWidth:
+            full_width_reg_value = value
+        elif ilreg.extend == ImplicitRegisterExtend.SignExtendToFullWidth:
+            full_width_reg_value = (
+                (value ^ ((1 << ilreg.size * 8) - 1)) -
+                ((1 << ilreg.size * 8) - 1) +
+                (1 << ilreg_full.size * 8)
+            )
+        else:
+            raise NotImplementedError
+
+        self.registers[ilreg.full_width_reg] = full_width_reg_value
+        # FIXME apply a mask here instead of doing it inside the instructions
+        return full_width_reg_value
+
+    def read(self, reg):
+        from binaryninja import Architecture
+
+        arch = Architecture[self.arch]
+        r = self._alias(str(reg))
+        if r in self.registers:
+            return self.registers[r]
+
+        reg_info = arch.regs.get(r)
+        if reg_info is None:
+            return None
+        full_reg_value = self.registers[reg_info.full_width_reg]
+        mask = (1 << reg_info.size * 8) - 1
+        reg_bits = mask << (reg_info.offset * 8)
+        reg_value = (full_reg_value & reg_bits) >> (reg_info.offset * 8)
+        # FIXME apply a mask here instead of doing it inside the instructions
+        return reg_value
+
+    @property
+    def all_registers(self):
+        return tuple(self.registers.keys() + self._aliases.keys())
+
+    @property
+    def canonical_registers(self):
+        return tuple(self.registers.keys())
+
+    def __contains__(self, register):
+        return register in self.all_registers
+
+    def get_pc(self):
+        if self.arch == 'x86':
+            return 'eip'
+        elif self.arch == 'x86_64':
+            return 'rip'
+        else:
+            raise NotImplementedError
+
+class BinjaOperand(Operand):
+    def __init__(self, cpu, parent_il, op, **kwargs):
+        self.llil = parent_il
+        if hasattr(op, "operands") and op.operands:
+            op.operands = [BinjaOperand(cpu, op, x) for x in op.operands]
+        super(BinjaOperand, self).__init__(cpu, op, **kwargs)
+
+    @property
+    def type(self):
+        from binaryninja import lowlevelil as il
+        type_map = {
+            il.ILRegister : 'register',
+            il.LowLevelILInstruction : 'instruction',
+            il.ILFlag : 'flag',
+        }
+
+        try:
+            t = type_map.get(type(self.op))
+        except Exception as e:
+            print "ERROR IN type lookup " + str(type(self.op))
+            raise e
+        return t
+
+    @property
+    def size(self):
+        # FIXME (does this assert need to be here? we could be reading memory)
```

Comment:
These discussion is better suited for github comments imho. It might not be needed; having it error is probably okay, but having an explicit assert is easier for catching the error while developing instruction semantics.